db/migrate/20230808140921_add_derived_remaining_hours_to_work_packages.rb

```diff
@@ -0,0 +1,69 @@
+# frozen_string_literal: true
+
+#--copyright
+# OpenProject is an open source project management software.
+# Copyright (C) 2012-2023 the OpenProject GmbH
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License version 3.
+#
+# OpenProject is a fork of ChiliProject, which is a fork of Redmine. The copyright follows:
+# Copyright (C) 2006-2013 Jean-Philippe Lang
+# Copyright (C) 2010-2013 the ChiliProject Team
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License
+# as published by the Free Software Foundation; either version 2
+# of the License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+#
+# See COPYRIGHT and LICENSE files for more details.
+#++
+
+class AddDerivedRemainingHoursToWorkPackages < ActiveRecord::Migration[7.0]
+  def change
+    add_column :work_packages, :derived_remaining_hours, :float
+    add_column :work_package_journals, :derived_remaining_hours, :float
+
+    reversible do |change|
+      change.up do
+        WorkPackage.transaction do
+          migrate_to_derived_remaining_hours!
+        end
+      end
+
+      change.down do
+        WorkPackage.transaction do
+          rollback_from_derived_remaining_hours!
+        end
+      end
+    end
+  end
+
+  def work_packages_to_update
+    leaf_node_ids = WorkPackage.leaves.pluck(:id)
+    WorkPackage.where.not(id: leaf_node_ids)
+  end
+
+  def migrate_to_derived_remaining_hours!
+    scope = work_packages_to_update
+              .where.not(remaining_hours: nil)
+    scope.update_all("derived_remaining_hours = remaining_hours, remaining_hours = NULL")
+  end
+
+  def journal_user = SystemUser.first
+
+  def rollback_from_derived_remaining_hours!
+    scope = work_packages_to_update
+              .where.not(derived_remaining_hours: nil)
+    scope.update_all("remaining_hours = derived_remaining_hours")
```

Comment:
the blocks are actually ran before any actual changes happen in the database, which is kinda neat for cases like this one.

Reply:
Nice.
And it probably runs in reverse only when rolling back, or the migration would not work (it could run the block which populates the columns before the column is added).